specs/spec.md

```diff
@@ -0,0 +1,329 @@
+
+* Peers may request blocks and headers from other peers, possibly in response to a `GOSSIP` message. 
+* Peers may request blocks repeatedly from the same peers.
+* Other peers may respond on a best effort basis with header and block data.
+* There is no SLA for responding. 
+
+##### Handshake
+
+###### `0x00` HELLO
+
+Upon discovering each other, nodes may exchange `HELLO` messages.
+
+Nodes may send `HELLO` to other peers when they exchange messages for the first time or when their state changes to let them know new blocks are available.
+
+Nodes can send `HELLO` messages to each other to exchange information on their status:
+
+```python
+{
+  'network_id': 'uint8' ## the ID of the network (1 for mainnet, and some predefined number for a testnet)
+  'chain_id': 'uint8' ## the ID of the chain (1 for ETH)
+  'latest_finalized_root': 'bytes32' ## the hash of the latest finalized root
+  'latest_finalized_epoch': 'uint64' ## the number of the latest finalized epoch
+  'best_root': 'bytes32' ## the hash of the best root this node can offer
+  'best_slot': 'uint64' ## the number of the best slot this node can offer
+}
+```
+
+Upon receiving a `HELLO` message, the node should reply with a `HELLO` message.
+
+###### `0x01` GOODBYE
+
+Nodes may signal to other nodes that they are going away by sending a `GOODBYE` message:
+
+```python
+{
+  'reason': 'uint64' ## an optional reason code up to the client
+}
+```
+
+The reason given is optional. Reason codes are up to each client and should not be trusted.
+
+Upon receiving a `GOODBYE` message, no response is necessary.
+
+##### `0x02` GET_STATUS
+
+Nodes may exchange metadata information using a `GET_STATUS` message.
+<!--is this necessary?-->
+A `GET_STATUS` request may be sent in response to receiving a `GOSSIP` message:
+
+```python
+{
+  'user_agent': 'bytes' ## the human readable name of the client, optionally with its version and other metadata
+  'timestamp': 'uint64' ## the current time of the node in milliseconds since epoch
+}
+```
+
+Any peer may provide information about their status and metadata to any other peer. Other peers may respond on a best effort basis, if at all.
+
+##### Block Headers
+
+###### `0x0A` GET_BLOCK_HEADERS
+
+Nodes may request block headers from other nodes using the `GET_BLOCK_HEADERS` message:
+
+<!--is this necessary?-->
+```python
+{
+  'start_root': 'bytes32' ## the root hash to start querying from OR
+  'start_slot': 'uint64' ## the slot number to start querying from
+  'max': 'uint64' ## the max number of elements to return
+  'skip': 'uint64' ## the number of elements apart to pick from
+  'direction': 'uint8' ## 0x01 is ascending, 0x00 is descending direction to query elements
+}
+```
+
+A `GET_BLOCK_HEADERS` request may be sent in response to receiving a `GOSSIP` message.
+
+###### `0x0B` BLOCK_HEADERS
+
+Nodes may provide block roots to other nodes using the `BLOCK_HEADERS` message, usually in response to a `GET_BLOCK_HEADERS` message:
```

Comment:
Why are the headers necessary as network objects? 
I think `GET_BLOCK_BODIES` and `BLOCK_BODIES` are sufficient enough. `BeaconBlock` contains header

Reply:
Same as LES. You sync headers fast, and bodies separately at a slower pace. Having them separated allows a number of other use cases such as light clients.